core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/ContainerProxy.scala

```diff
@@ -479,22 +497,37 @@ class ContainerProxy(factory: (TransactionId,
 
   /** Either process runbuffer or signal parent to send work; return true if runbuffer is being processed */
   def requestWork(newData: WarmedData): Boolean = {
-    //if there is concurrency capacity, process runbuffer, or signal NeedWork
+    //if there is concurrency capacity, process runbuffer, signal NeedWork, or both
     if (activeCount < newData.action.limits.concurrency.maxConcurrent) {
-      runBuffer.dequeueOption match {
-        case Some((run, q)) =>
-          runBuffer = q
-          self ! run
-          true
-        case _ =>
-          context.parent ! NeedWork(newData)
-          false
+      if (runBuffer.length > 0) {
+        processBuffer(newData.action, newData)
+        true
+      } else {
+        context.parent ! NeedWork(newData)
+        bufferProcessing //true in case buffer is still in process
       }
     } else {
       false
     }
   }
 
+  /** Process buffered items up to the capacity of action concurrency config */
+  def processBuffer(action: ExecutableWhiskAction, newData: ContainerData) = {
+    //send as many buffered as possible
+    var available = action.limits.concurrency.maxConcurrent - activeCount
+    logging.info(this, s"resending up to ${available} from ${runBuffer.length} buffered jobs")
+    while (available > 0 && runBuffer.length > 0) {
```

Comment:
Small nit.
How about replacing `var` and `while` with something similar to this?

```scala
    val available = action.limits.concurrency.maxConcurrent - activeCount
    logging.info(this, s"resending up to ${available} from ${runBuffer.length} buffered jobs")
    1 to available foreach { _ =>
      runBuffer.dequeueOption match {
        case Some((run, q)) =>
          self ! run
          bufferProcessing = true
          runBuffer = q
        case _ =>
          context.parent ! NeedWork(newData)
      }
    }
```

Reply:
Yes this is nicer!

Reply:
nice @style95 👏